<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
  <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
  <meta http-equiv="Content-Style-Type" content="text/css" />
  <meta name="generator" content="pandoc" />
  <title></title>
  <style type="text/css">code{white-space: pre;}</style>
  <style type="text/css">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
code > span.kw { color: #007020; font-weight: bold; } /* Keyword */
code > span.dt { color: #902000; } /* DataType */
code > span.dv { color: #40a070; } /* DecVal */
code > span.bn { color: #40a070; } /* BaseN */
code > span.fl { color: #40a070; } /* Float */
code > span.ch { color: #4070a0; } /* Char */
code > span.st { color: #4070a0; } /* String */
code > span.co { color: #60a0b0; font-style: italic; } /* Comment */
code > span.ot { color: #007020; } /* Other */
code > span.al { color: #ff0000; font-weight: bold; } /* Alert */
code > span.fu { color: #06287e; } /* Function */
code > span.er { color: #ff0000; font-weight: bold; } /* Error */
code > span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
code > span.cn { color: #880000; } /* Constant */
code > span.sc { color: #4070a0; } /* SpecialChar */
code > span.vs { color: #4070a0; } /* VerbatimString */
code > span.ss { color: #bb6688; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { color: #19177c; } /* Variable */
code > span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code > span.op { color: #666666; } /* Operator */
code > span.bu { } /* BuiltIn */
code > span.ex { } /* Extension */
code > span.pp { color: #bc7a00; } /* Preprocessor */
code > span.at { color: #7d9029; } /* Attribute */
code > span.do { color: #ba2121; font-style: italic; } /* Documentation */
code > span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code > span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code > span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
  </style>
  <link rel="stylesheet" href="github-markdown.css" type="text/css" />
</head>
<body>
<div id="TOC">
<ul>
<li><a href="#script-information"><span class="toc-section-number">1</span> Script information</a><ul>
<li><a href="#year"><span class="toc-section-number">1.1</span> Year</a></li>
<li><a href="#authors"><span class="toc-section-number">1.2</span> Authors</a></li>
<li><a href="#affiliations"><span class="toc-section-number">1.3</span> Affiliations</a></li>
</ul></li>
<li><a href="#description"><span class="toc-section-number">2</span> Description</a></li>
<li><a href="#set-use-utilities"><span class="toc-section-number">3</span> Set `use utilities</a></li>
<li><a href="#set-variables"><span class="toc-section-number">4</span> Set variables</a></li>
<li><a href="#displays-help"><span class="toc-section-number">5</span> Displays help</a></li>
</ul>
</div>
<div class="sourceCode"><pre class="sourceCode perl"><code class="sourceCode perl"><span class="kw">#!/usr/bin/env perl</span></code></pre></div>
<h1 id="script-information"><span class="header-section-number">1</span> Script information</h1>
<h2 id="year"><span class="header-section-number">1.1</span> Year</h2>
<ul>
<li>2016</li>
</ul>
<h2 id="authors"><span class="header-section-number">1.2</span> Authors</h2>
<ul>
<li>Bruno Contreras-Moreira (1) -Pablo Vinuesa (2):</li>
</ul>
<h2 id="affiliations"><span class="header-section-number">1.3</span> Affiliations</h2>
<ul>
<li>1: <a href="http://www.eead.csic.es/compbio">Laboratory of Computational Biology, EEAD/CSIC, Spain)</a></li>
<li>2: <a href="http://www.ccg.unam.mx/~vinuesa">Center for Genomic Sciences, UNAM, Mexico)</a></li>
</ul>
<h1 id="description"><span class="header-section-number">2</span> Description</h1>
<p>This script produces a multiple alignment view of <strong>BLAST</strong> locally-aligned sequences clustered by `get_homologues[-est].pl It does not necessarily conserved the original sequence order</p>
<p>Uses third-party software: MVIEW (https://github.com/desmid/mview) Brown NP, Leroy C, Sander C (1998) MView: A Web compatible database search or multiple alignment viewer. Bioinformatics. 14 (4):380-381.</p>
<div class="sourceCode"><pre class="sourceCode perl"><code class="sourceCode perl"><span class="dt">$|</span>=<span class="dv">1</span>;</code></pre></div>
<h1 id="set-use-utilities"><span class="header-section-number">3</span> Set `use utilities</h1>
<div class="sourceCode"><pre class="sourceCode perl"><code class="sourceCode perl"><span class="fu">use</span> <span class="kw">strict</span>;
<span class="fu">use</span> <span class="fu">Getopt::Std</span>;
<span class="fu">use</span> FileHandle;
<span class="fu">use</span> <span class="fu">File::Temp</span> <span class="kw">qw/</span>tempfile<span class="kw">/</span>;
<span class="fu">use</span> FindBin <span class="kw">&#39;</span><span class="st">$Bin</span><span class="kw">&#39;</span>;
<span class="fu">use</span> lib <span class="kw">&quot;</span><span class="dt">$Bin</span><span class="st">/lib</span><span class="kw">&quot;</span>;
<span class="fu">use</span> lib <span class="kw">&quot;</span><span class="dt">$Bin</span><span class="st">/lib/bioperl-1.5.2_102/</span><span class="kw">&quot;</span>;
<span class="fu">use</span> <span class="fu">Bio::Seq</span>;
<span class="fu">use</span> <span class="fu">Bio::SeqUtils</span>;
<span class="fu">use</span> phyTools;
<span class="fu">use</span> marfil_homology;</code></pre></div>
<h1 id="set-variables"><span class="header-section-number">4</span> Set variables</h1>
<div class="sourceCode"><pre class="sourceCode perl"><code class="sourceCode perl"><span class="kw">my</span> <span class="dt">@FEATURES2CHECK</span> = (<span class="kw">&#39;</span><span class="st">EXE_BLASTP</span><span class="kw">&#39;</span>,<span class="kw">&#39;</span><span class="st">EXE_BLASTN</span><span class="kw">&#39;</span>,<span class="kw">&#39;</span><span class="st">EXE_FORMATDB</span><span class="kw">&#39;</span>,<span class="kw">&#39;</span><span class="st">EXE_MVIEW</span><span class="kw">&#39;</span>,<span class="kw">&#39;</span><span class="st">EXE_HMMPFAM</span><span class="kw">&#39;</span>);
<span class="kw">my</span> <span class="dt">$DEFBLASTNTASK</span>  = <span class="kw">&#39;</span><span class="st">megablast</span><span class="kw">&#39;</span>;
<span class="kw">my</span> <span class="dt">$DEFEVALUE</span>      = <span class="dv">10</span>; <span class="co"># default BLAST E-value</span>
<span class="kw">my</span> <span class="dt">$MINBLUNTBLOCK</span>  = <span class="dv">100</span>; <span class="co"># min alignment width with blunt ends</span>
<span class="kw">my</span> (<span class="dt">$INP_nucleotides</span>,<span class="dt">$INP_blunt</span>,<span class="dt">$do_PFAM</span>,<span class="dt">$INP_clusterfile</span>,<span class="dt">$INP_outfile</span>,<span class="dt">%opts</span>) = (<span class="dv">1</span>,<span class="dv">0</span>,<span class="dv">0</span>,<span class="kw">&#39;&#39;</span>,<span class="kw">&#39;&#39;</span>);
getopts(<span class="kw">&#39;</span><span class="st">hDbPf:o:</span><span class="kw">&#39;</span>, \<span class="dt">%opts</span>);</code></pre></div>
<h1 id="displays-help"><span class="header-section-number">5</span> Displays help</h1>
<div class="sourceCode"><pre class="sourceCode perl"><code class="sourceCode perl"><span class="kw">if</span>((<span class="dt">$opts</span>{<span class="kw">&#39;</span><span class="st">h</span><span class="kw">&#39;</span>})||(<span class="fu">scalar</span>(<span class="fu">keys</span>(<span class="dt">%opts</span>))==<span class="dv">0</span>))
{
  <span class="fu">print</span>   <span class="kw">&quot;</span><span class="ch">\n</span><span class="st">usage: </span><span class="dt">$0</span><span class="st"> [options]</span><span class="ch">\n\n</span><span class="kw">&quot;</span>;
  <span class="fu">print</span>   <span class="kw">&quot;</span><span class="st">-h this message</span><span class="ch">\n</span><span class="kw">&quot;</span>;
  <span class="fu">print</span>   <span class="kw">&quot;</span><span class="st">-f input cluster FASTA file            (expects nucleotides)</span><span class="ch">\n</span><span class="kw">&quot;</span>;        
  <span class="fu">print</span>   <span class="kw">&quot;</span><span class="st">-o output alignment file               (optional, produces FASTA format)</span><span class="ch">\n</span><span class="kw">&quot;</span>;
  <span class="fu">print</span>   <span class="kw">&quot;</span><span class="st">-P sequences are peptides              (optional)</span><span class="ch">\n</span><span class="kw">&quot;</span>;
  <span class="fu">print</span>   <span class="kw">&quot;</span><span class="st">-b blunt alignment borders             (optional, otherwise all bits aligned by BLAST are shown)</span><span class="ch">\n</span><span class="kw">&quot;</span>;
  <span class="fu">print</span>   <span class="kw">&quot;</span><span class="st">-D match Pfam domains                  (optional, based on six-frame translation of longest seq)</span><span class="ch">\n</span><span class="kw">&quot;</span>;
  <span class="fu">exit</span>(<span class="dv">0</span>);
}
<span class="fu">print</span> STDERR check_installed_features(<span class="dt">@FEATURES2CHECK</span>);
<span class="kw">if</span>(<span class="fu">defined</span>(<span class="dt">$opts</span>{<span class="kw">&#39;</span><span class="st">f</span><span class="kw">&#39;</span>})){  <span class="dt">$INP_clusterfile</span> = <span class="dt">$opts</span>{<span class="kw">&#39;</span><span class="st">f</span><span class="kw">&#39;</span>}; }
<span class="kw">else</span>{ <span class="fu">die</span> <span class="kw">&quot;</span><span class="st"># EXIT : need -f cluster FASTA file</span><span class="ch">\n</span><span class="kw">&quot;</span>; }
<span class="kw">if</span>(<span class="fu">defined</span>(<span class="dt">$opts</span>{<span class="kw">&#39;</span><span class="st">o</span><span class="kw">&#39;</span>})){  <span class="dt">$INP_outfile</span> = <span class="dt">$opts</span>{<span class="kw">&#39;</span><span class="st">o</span><span class="kw">&#39;</span>}; }
<span class="kw">if</span>(<span class="fu">defined</span>(<span class="dt">$opts</span>{<span class="kw">&#39;</span><span class="st">P</span><span class="kw">&#39;</span>})){  <span class="dt">$INP_nucleotides</span> = <span class="dv">0</span> }
<span class="kw">if</span>(<span class="fu">defined</span>(<span class="dt">$opts</span>{<span class="kw">&#39;</span><span class="st">b</span><span class="kw">&#39;</span>}))
{ 
    <span class="dt">$INP_blunt</span> = <span class="dt">$MINBLUNTBLOCK</span>;
}
<span class="kw">if</span>(<span class="fu">defined</span>(<span class="dt">$opts</span>{<span class="kw">&#39;</span><span class="st">D</span><span class="kw">&#39;</span>}))
{
  <span class="kw">if</span>(feature_is_installed(<span class="kw">&#39;</span><span class="st">PFAM</span><span class="kw">&#39;</span>))
  {
    <span class="dt">$do_PFAM</span> = <span class="dv">1</span>;
  }
  <span class="kw">else</span>{ warn_missing_soft(<span class="kw">&#39;</span><span class="st">PFAM</span><span class="kw">&#39;</span>) }
}
<span class="fu">printf</span>(STDERR <span class="kw">&quot;</span><span class="ch">\n</span><span class="st"># </span><span class="dt">%s</span><span class="st"> -f </span><span class="dt">%s</span><span class="st"> -o </span><span class="dt">%s</span><span class="st"> -P </span><span class="dt">%s</span><span class="st"> -b </span><span class="dt">%s</span><span class="st"> -D </span><span class="dt">%s</span><span class="ch">\n</span><span class="kw">&quot;</span>,
  <span class="dt">$0</span>,<span class="dt">$INP_clusterfile</span>,<span class="dt">$INP_outfile</span>,<span class="dt">$INP_nucleotides</span>,<span class="dt">$INP_blunt</span>,<span class="dt">$do_PFAM</span>);
<span class="co">##########################################################################</span>
<span class="kw">my</span> (<span class="dt">$maxlength</span>,<span class="dt">$longest_seq</span>,<span class="dt">$length</span>,<span class="dt">$start</span>,<span class="dt">$end</span>,<span class="dt">$seq</span>,<span class="dt">$fh</span>,<span class="dt">$seqname</span>) = (<span class="dv">0</span>,-<span class="dv">1</span>);
<span class="kw">my</span> (<span class="dt">$command</span>,<span class="dt">$Pfam_string</span>,<span class="dt">$Pfam_full</span>,<span class="dt">@trash</span>,<span class="dt">%short_names</span>);
<span class="co">## parse input cluster and get sequences</span>
<span class="kw">my</span> <span class="dt">$cluster_ref</span> = read_FASTA_file_array(<span class="dt">$INP_clusterfile</span>,<span class="dt">$INP_nucleotides</span>);
<span class="kw">foreach</span> <span class="dt">$seq</span> (<span class="dv">0</span> .. <span class="dt">$#</span>{<span class="dt">$cluster_ref</span>})
{
  <span class="co"># shorten name</span>
  <span class="kw">if</span>(<span class="dt">$cluster_ref</span>-&gt;[<span class="dt">$seq</span>][NAME] =~ <span class="kw">m/</span><span class="ch">^(</span><span class="bn">\S</span><span class="ch">+)</span><span class="ot">.</span><span class="ch">*?(</span><span class="ot">\[</span><span class="bn">\S</span><span class="ch">+?</span><span class="ot">\]</span><span class="ch">)$</span><span class="kw">/</span> || 
    <span class="dt">$cluster_ref</span>-&gt;[<span class="dt">$seq</span>][NAME] =~ <span class="kw">m/</span><span class="ch">^(</span><span class="bn">\S</span><span class="ch">+)</span><span class="ot">.</span><span class="ch">*?(</span><span class="ot">\[</span><span class="bn">\S</span><span class="ch">+?</span><span class="ot">\]</span><span class="ch">)</span><span class="bn">\s</span><span class="ot">\|</span><span class="kw">/</span>)
  {
    <span class="dt">$seqname</span> = <span class="kw">&quot;</span><span class="dt">$1</span><span class="st">\_</span><span class="dt">$2</span><span class="kw">&quot;</span>; 
  }
  <span class="kw">else</span>
  {
    <span class="dt">$seqname</span> = (<span class="fu">split</span>(<span class="kw">/</span><span class="bn">\s</span><span class="ch">+</span><span class="kw">/</span>,<span class="dt">$cluster_ref</span>-&gt;[<span class="dt">$seq</span>][NAME]))[<span class="dv">0</span>];
  }
  <span class="dt">$short_names</span>{<span class="dt">$seq</span>} = <span class="dt">$seqname</span>;
  <span class="co"># in case of missing sequences</span>
  <span class="kw">if</span>(!<span class="dt">$cluster_ref</span>-&gt;[<span class="dt">$seq</span>][SEQ])
  { 
    <span class="fu">print</span> STDERR <span class="kw">&quot;</span><span class="st"># skip </span><span class="dt">$seq</span><span class="st"> : </span><span class="dt">$cluster_ref</span>-&gt;<span class="st">[</span><span class="dt">$seq</span><span class="st">][NAME] due to missing sequence</span><span class="ch">\n</span><span class="kw">&quot;</span>;
    <span class="kw">next</span>;
  }  
  <span class="co"># find longest aligned sequence to be used as reference | aligned:1-296 (296)</span>
  <span class="kw">if</span>(<span class="dt">$cluster_ref</span>-&gt;[<span class="dt">$seq</span>][NAME] =~ <span class="kw">/</span><span class="ot"> \| aligned:</span><span class="ch">(</span><span class="bn">\d</span><span class="ch">+)</span><span class="ot">-</span><span class="ch">(</span><span class="bn">\d</span><span class="ch">+)</span><span class="ot"> </span><span class="kw">/</span>)
  { 
    (<span class="dt">$start</span>,<span class="dt">$end</span>) = (<span class="dt">$1</span>,<span class="dt">$2</span>);
    <span class="dt">$length</span> = (<span class="dt">$end</span>-<span class="dt">$start</span>)+<span class="dv">1</span>;
  }
  <span class="kw">else</span>
  {
    <span class="dt">$length</span> = <span class="fu">length</span>(<span class="dt">$cluster_ref</span>-&gt;[<span class="dt">$seq</span>][SEQ]);
  }
  
  <span class="co"># find longest [aligned] sequence</span>
  <span class="kw">if</span>(<span class="dt">$maxlength</span> == <span class="dv">0</span> || <span class="dt">$length</span> &gt; <span class="dt">$maxlength</span>)
  {
    <span class="dt">$maxlength</span> = <span class="dt">$length</span>;
    <span class="dt">$longest_seq</span> = <span class="dt">$seq</span>;
  }
}
<span class="co">#printf(STDERR &quot;\n# total   sequences: %d longest: %d\n&quot;,$#{$cluster_ref}+1,$longest_seq+1); </span>
<span class="fu">printf</span>(STDERR <span class="kw">&quot;</span><span class="ch">\n</span><span class="st"># total   sequences: </span><span class="dt">%d</span><span class="ch">\n</span><span class="kw">&quot;</span>,<span class="dt">$#</span>{<span class="dt">$cluster_ref</span>}+<span class="dv">1</span>); 
<span class="kw">if</span>(<span class="dt">$#</span>{<span class="dt">$cluster_ref</span>} == -<span class="dv">1</span>)
{
  <span class="fu">warn</span> <span class="kw">&quot;</span><span class="st"># Cannot read input sequences, exit. Please set -P if using a peptide cluster.</span><span class="ch">\n</span><span class="kw">&quot;</span>;
  <span class="fu">exit</span>;
}
<span class="co">## Pfam-annotate longest sequence if required</span>
<span class="kw">if</span>(<span class="dt">$do_PFAM</span>)
{
  <span class="kw">my</span> (<span class="dt">$fhpf</span>,<span class="dt">$filenamepf</span>) = tempfile(UNLINK =&gt; <span class="dv">1</span>); <span class="co"># Pfam input </span>
  <span class="kw">my</span> (<span class="dt">$fhpo</span>,<span class="dt">$filenamepo</span>) = tempfile(UNLINK =&gt; <span class="dv">1</span>); <span class="co"># Pfam output</span>
  <span class="kw">my</span> (<span class="dt">$fhpr</span>,<span class="dt">$filenamepr</span>) = tempfile(UNLINK =&gt; <span class="dv">1</span>); <span class="co"># Pfam report</span>
  <span class="kw">if</span>(<span class="dt">$INP_nucleotides</span>)
  {
    <span class="co"># translate to 6 frames </span>
    <span class="co"># https://cryptogenomicon.org/2011/05/27/hmmscan-vs-hmmsearch-speed-the-numerology/</span>
    <span class="kw">my</span> <span class="dt">$seqobj</span> = <span class="fu">Bio::Seq</span>-&gt;new (-seq=&gt;<span class="dt">$cluster_ref</span>-&gt;[<span class="dt">$longest_seq</span>][SEQ],-alphabet=&gt;<span class="kw">&#39;</span><span class="st">DNA</span><span class="kw">&#39;</span>);
    <span class="kw">my</span> <span class="dt">$utils</span> = new <span class="fu">Bio::SeqUtils</span>;
    <span class="kw">my</span> <span class="dt">@seqs</span> = <span class="dt">$utils</span>-&gt;<span class="dt">translate_6frames</span>(<span class="dt">$seqobj</span>);
    <span class="kw">foreach</span> <span class="dt">$seq</span> (<span class="dv">0</span> .. <span class="dt">$#seqs</span>)
    {
      <span class="fu">print</span> <span class="dt">$fhpf</span> <span class="kw">&quot;</span><span class="st">&gt;</span><span class="dt">$seq</span><span class="ch">\n</span><span class="dt">$seqs</span><span class="st">[</span><span class="dt">$seq</span><span class="st">]-&gt;{&#39;primary_seq&#39;}{&#39;seq&#39;}</span><span class="ch">\n</span><span class="kw">&quot;</span>;
    }
  }
  <span class="kw">else</span>
  {
    <span class="fu">print</span> <span class="dt">$fhpf</span> <span class="kw">&quot;</span><span class="st">&gt;1</span><span class="ch">\n</span><span class="dt">$cluster_ref</span>-&gt;<span class="st">[</span><span class="dt">$longest_seq</span><span class="st">][SEQ]</span><span class="ch">\n</span><span class="kw">&quot;</span>;
  }  
  
  <span class="fu">close</span>(<span class="dt">$fhpf</span>);
  <span class="fu">close</span>(<span class="dt">$fhpo</span>);
  <span class="fu">close</span>(<span class="dt">$fhpr</span>);
  
  <span class="dt">$command</span> = format_HMMPFAM_command().<span class="kw">&quot;</span><span class="st"> </span><span class="dt">$filenamepf</span><span class="st"> &gt; </span><span class="dt">$filenamepo</span><span class="st"> </span><span class="kw">&quot;</span>;
  <span class="fu">system</span>(<span class="dt">$command</span>);
  <span class="kw">if</span>(<span class="dt">$?</span> != <span class="dv">0</span>)
  {
    <span class="fu">die</span> <span class="kw">&quot;</span><span class="st"># EXIT: failed while running Pfam search (</span><span class="dt">$command</span><span class="st">)</span><span class="ch">\n</span><span class="kw">&quot;</span>;
  }<span class="co">#else{ system(&quot;cat $filenamepo&quot;); } # testing</span>
  
  pfam_parse(<span class="dt">$filenamepr</span>,(<span class="dt">$filenamepo</span>));
  
  <span class="fu">open</span>(PFAMREPORT,<span class="dt">$filenamepr</span>);
  <span class="kw">while</span>(<span class="kw">&lt;PFAMREPORT&gt;</span>)
  {
    <span class="kw">if</span>(<span class="kw">/</span><span class="ch">^</span><span class="bn">\d</span><span class="ch">+</span><span class="ot">\t</span><span class="ch">(</span><span class="ot">PF</span><span class="bn">\S</span><span class="ch">+)</span><span class="ot">\t</span><span class="ch">(</span><span class="ot">.</span><span class="ch">*?)</span><span class="bn">\n</span><span class="kw">/</span>)
    {
      <span class="dt">$Pfam_string</span> .= <span class="dt">$1</span>;
      <span class="dt">$Pfam_full</span>   .= <span class="dt">$2</span>;
    }  
  }
  <span class="fu">close</span>(PFAMREPORT);
  
  <span class="fu">printf</span>(STDERR <span class="kw">&quot;</span><span class="ch">\n</span><span class="st"># Pfam domains: </span><span class="dt">%s</span><span class="ch">\n</span><span class="kw">&quot;</span>,<span class="dt">$Pfam_string</span>); 
  <span class="fu">printf</span>(STDERR <span class="kw">&quot;</span><span class="st"># Pfam annotation: </span><span class="dt">%s</span><span class="ch">\n</span><span class="kw">&quot;</span>,<span class="dt">$Pfam_full</span>); 
}  
  
<span class="co">## align cluster sequences</span>
<span class="co"># temp files, removed automatically on exit</span>
<span class="kw">my</span> (<span class="dt">$fhdb</span>,<span class="dt">$filenamedb</span>) = tempfile(UNLINK =&gt; <span class="dv">1</span>); <span class="co"># blast DB</span>
<span class="kw">my</span> (<span class="dt">$fhq</span>,<span class="dt">$filenameq</span>)   = tempfile(UNLINK =&gt; <span class="dv">1</span>); <span class="co"># blast query</span>
<span class="kw">my</span> (<span class="dt">$fhb</span>,<span class="dt">$filenameb</span>)   = tempfile(UNLINK =&gt; <span class="dv">1</span>); <span class="co"># blast output</span>
<span class="kw">my</span> (<span class="dt">$fha</span>,<span class="dt">$filenamea</span>)   = tempfile(UNLINK =&gt; <span class="dv">1</span>); <span class="co"># out alignment file</span>
<span class="kw">foreach</span> <span class="dt">$seq</span> (<span class="dv">0</span> .. <span class="dt">$#</span>{<span class="dt">$cluster_ref</span>})
{
  <span class="kw">if</span>(<span class="dt">$seq</span> == <span class="dt">$longest_seq</span>){ <span class="dt">$fh</span> = <span class="dt">$fhq</span> }
  <span class="kw">else</span>{ <span class="dt">$fh</span> = <span class="dt">$fhdb</span> }
  
  <span class="fu">print</span> <span class="dt">$fh</span> <span class="kw">&quot;</span><span class="st">&gt;</span><span class="dt">$short_names</span><span class="st">{</span><span class="dt">$seq</span><span class="st">}</span><span class="ch">\n</span><span class="dt">$cluster_ref</span>-&gt;<span class="st">[</span><span class="dt">$seq</span><span class="st">][SEQ]</span><span class="ch">\n</span><span class="kw">&quot;</span>;
}
<span class="fu">close</span>(<span class="dt">$fha</span>);
<span class="fu">close</span>(<span class="dt">$fhb</span>);
<span class="fu">close</span>(<span class="dt">$fhq</span>);
<span class="fu">close</span>(<span class="dt">$fhdb</span>);
<span class="kw">if</span>(<span class="dt">$INP_nucleotides</span>)
{
  executeFORMATDB(<span class="dt">$filenamedb</span>,<span class="dv">1</span>,<span class="dv">1</span>); 
  <span class="dt">$command</span> = format_BLASTN_command_aligns(<span class="dt">$filenameq</span>,<span class="dt">$filenameb</span>,<span class="dt">$filenamedb</span>,<span class="dt">$DEFEVALUE</span>,<span class="dt">$DEFBLASTNTASK</span>);
  <span class="fu">push</span>(<span class="dt">@trash</span>,<span class="dt">$fhdb</span>.<span class="kw">&#39;</span><span class="st">.nsq</span><span class="kw">&#39;</span>, <span class="dt">$fhdb</span>.<span class="kw">&#39;</span><span class="st">.nin</span><span class="kw">&#39;</span>, <span class="dt">$fhdb</span>.<span class="kw">&#39;</span><span class="st">.nhr</span><span class="kw">&#39;</span>);
}
<span class="kw">else</span>
{
  executeFORMATDB(<span class="dt">$filenamedb</span>,<span class="dv">0</span>,<span class="dv">1</span>);
  <span class="dt">$command</span> = format_BLASTP_command_aligns(<span class="dt">$filenameq</span>,<span class="dt">$filenameb</span>,<span class="dt">$filenamedb</span>,<span class="dt">$DEFEVALUE</span>);
  <span class="fu">push</span>(<span class="dt">@trash</span>,<span class="dt">$fhdb</span>.<span class="kw">&#39;</span><span class="st">.psq</span><span class="kw">&#39;</span>, <span class="dt">$fhdb</span>.<span class="kw">&#39;</span><span class="st">.pin</span><span class="kw">&#39;</span>, <span class="dt">$fhdb</span>.<span class="kw">&#39;</span><span class="st">.phr</span><span class="kw">&#39;</span>);
}
<span class="fu">system</span>(<span class="dt">$command</span>);
<span class="kw">if</span>(<span class="dt">$?</span> != <span class="dv">0</span>)
{
  <span class="fu">die</span> <span class="kw">&quot;</span><span class="st"># EXIT: failed while running BLAST search (</span><span class="dt">$command</span><span class="st">)</span><span class="ch">\n</span><span class="kw">&quot;</span>;
}
<span class="kw">else</span>
{
  <span class="co"># clean BLAST tmp files</span>
  <span class="fu">unlink</span>(<span class="dt">@trash</span>);
}
<span class="co">## format raw alignment  #-width 60 </span>
<span class="dt">$command</span> = <span class="kw">&quot;</span><span class="dt">$ENV</span><span class="st">{&#39;EXE_MVIEW&#39;} -in blast -out fasta </span><span class="dt">$filenameb</span><span class="st"> &gt; </span><span class="dt">$filenamea</span><span class="kw">&quot;</span>; 
<span class="fu">system</span>(<span class="dt">$command</span>); 
<span class="co">## extract SNPs and parsimony-informative sites, and trim alignment if required</span>
<span class="kw">my</span> <span class="dt">$align_ref</span> = check_variants_FASTA_alignment(<span class="dt">$filenamea</span>,!<span class="dt">$INP_nucleotides</span>,<span class="dt">$INP_blunt</span>);
  
<span class="fu">printf</span>(STDERR <span class="kw">&quot;</span><span class="st"># aligned sequences: </span><span class="dt">%d</span><span class="st"> width:   </span><span class="dt">%d</span><span class="ch">\n</span><span class="kw">&quot;</span>,<span class="dt">$#</span>{<span class="dt">$align_ref</span>}+<span class="dv">1</span>,<span class="fu">length</span>(<span class="dt">$align_ref</span>-&gt;[<span class="dv">0</span>][SEQ]));  
  
<span class="co">## print final alignment </span>
<span class="kw">if</span>(<span class="dt">$INP_outfile</span>)
{
  <span class="dt">$fh</span> = FileHandle-&gt;new(<span class="kw">&quot;</span><span class="st">&gt;</span><span class="dt">$INP_outfile</span><span class="kw">&quot;</span>);
  <span class="kw">if</span>(!<span class="fu">defined</span>(<span class="dt">$fh</span>))
  {
    <span class="fu">die</span> <span class="kw">&quot;</span><span class="st"># cannot create output file </span><span class="dt">$INP_outfile</span><span class="ch">\n</span><span class="kw">&quot;</span>;
  }
}
<span class="kw">else</span>{ <span class="dt">$fh</span> = <span class="dt">*STDOUT</span> }
<span class="kw">foreach</span> <span class="dt">$seq</span> (<span class="dv">0</span> .. <span class="dt">$#</span>{<span class="dt">$align_ref</span>})
{
  <span class="fu">print</span> <span class="dt">$fh</span> <span class="kw">&quot;</span><span class="st">&gt;</span><span class="dt">$align_ref</span>-&gt;<span class="st">[</span><span class="dt">$seq</span><span class="st">][NAME]</span><span class="kw">&quot;</span>;
  
  <span class="kw">if</span>(<span class="dt">$do_PFAM</span>){ <span class="fu">print</span> <span class="dt">$fh</span> <span class="kw">&quot;</span><span class="st"> Pfam:</span><span class="dt">$Pfam_full</span><span class="st">(</span><span class="dt">$Pfam_string</span><span class="st">)</span><span class="kw">&quot;</span> }
  
  <span class="fu">print</span> <span class="dt">$fh</span> <span class="kw">&quot;</span><span class="ch">\n</span><span class="dt">$align_ref</span>-&gt;<span class="st">[</span><span class="dt">$seq</span><span class="st">][SEQ]</span><span class="ch">\n</span><span class="kw">&quot;</span>;
} 
<span class="kw">if</span>(<span class="dt">$INP_outfile</span> &amp;&amp; <span class="dt">$#</span>{<span class="dt">$align_ref</span>})
{
  <span class="fu">print</span> STDERR <span class="kw">&quot;</span><span class="st"># alignment file: </span><span class="dt">$INP_outfile</span><span class="ch">\n</span><span class="kw">&quot;</span>; 
}  </code></pre></div>
</body>
</html>
